x/gamm/keeper/pool_service_test.go

```diff
@@ -203,241 +151,347 @@ func (suite *KeeperTestSuite) TestCreateBalancerPool() {
 		{
-			fn: func(poolId uint64) {
-				keeper := suite.App.GAMMKeeper
-
-				_, err := keeper.ExitPool(suite.Ctx, suite.TestAccs[0], poolId, sdk.NewInt(-1), sdk.Coins{})
-				suite.Require().Error(err, "can't join the pool with requesting negative share amount")
-			},
+			name:         "attempt exit pool requesting negative share amount",
+			txSender:     suite.TestAccs[0],
+			sharesIn:     sdk.NewInt(-1),
+			tokenOutMins: sdk.Coins{},
+			expectPass:   false,
 		},
 		{
-			fn: func(poolId uint64) {
-				keeper := suite.App.GAMMKeeper
-
-				// Test the "tokenOutMins"
-				// In this case, to refund the 50000000 amount of share token, the foo, bar token are expected to be refunded as 5000 amounts.
-				_, err := keeper.ExitPool(suite.Ctx, suite.TestAccs[0], poolId, types.OneShare.MulRaw(50), sdk.Coins{
-					sdk.NewCoin("foo", sdk.NewInt(5001)),
-				})
-				suite.Require().Error(err)
+			name:     "attempt exit pool with tokenOutMins above actual output",
+			txSender: suite.TestAccs[0],
+			sharesIn: types.OneShare.MulRaw(50),
+			tokenOutMins: sdk.Coins{
+				sdk.NewCoin("foo", sdk.NewInt(5001)),
 			},
+			expectPass: false,
 		},
 		{
-			fn: func(poolId uint64) {
-				keeper := suite.App.GAMMKeeper
-
-				// Test the "tokenOutMins"
-				// In this case, to refund the 50000000 amount of share token, the foo, bar token are expected to be refunded as 5000 amounts.
-				_, err := keeper.ExitPool(suite.Ctx, suite.TestAccs[0], poolId, types.OneShare.MulRaw(50), sdk.Coins{
-					sdk.NewCoin("foo", sdk.NewInt(5000)),
-				})
-				suite.Require().NoError(err)
+			name:     "attempt exit pool requesting tokenOutMins at exactly the actual output",
+			txSender: suite.TestAccs[0],
+			sharesIn: types.OneShare.MulRaw(50),
+			tokenOutMins: sdk.Coins{
+				sdk.NewCoin("foo", sdk.NewInt(5000)),
 			},
+			expectPass: true,
 		},
 	}
 
 	for _, test := range tests {
 		suite.SetupTest()
 
-		// Mint some assets to the accounts.
-		for _, acc := range suite.TestAccs {
-			suite.FundAcc(acc, defaultAcctFunds)
+		keeper := suite.App.GAMMKeeper
 
-			// Create the pool at first
-			msg := balancer.NewMsgCreateBalancerPool(suite.TestAccs[0], balancer.PoolParams{
-				SwapFee: sdk.NewDecWithPrec(1, 2),
-				ExitFee: sdk.NewDec(0),
-			}, defaultPoolAssets, defaultFutureGovernor)
-			poolId, err := suite.App.GAMMKeeper.CreatePool(suite.Ctx, msg)
-			suite.Require().NoError(err)
+		// Mint assets to the pool creator
+		suite.FundAcc(test.txSender, defaultAcctFunds)
 
-			test.fn(poolId)
+		// Create the pool at first
+		msg := balancer.NewMsgCreateBalancerPool(test.txSender, balancer.PoolParams{
+			SwapFee: sdk.NewDecWithPrec(1, 2),
+			ExitFee: sdk.NewDec(0),
+		}, defaultPoolAssets, defaultFutureGovernor)
+		poolId, err := suite.App.GAMMKeeper.CreatePool(suite.Ctx, msg)
+
+		// If we are testing insufficient pool share balances, switch sender to empty account
+		if test.txSender.String() == emptySender.String() {
+			test.txSender = suite.TestAccs[2]
```

Comment:
What's ther reason for changing emptySender from suite.TestAccs[1] to suite.TestAccs[2]? From my understanding, didn't we reserve suite.TestAccs[1] for the empty sender?

Reply:
I think this was a hacky approach to begin with – I refactored in recent commits to have an `emptySender` field for tests that need it.

For this specific case, we cannot use the same account we used to create the pool since that one has pool shares in it. If we create the pool with a completely independent account, then we would need to find and transfer the gamm shares for every other test, which felt like a messier approach than just switching the tx sender to an empty account when we want to test insufficient shares

Reply:
ah makes sense, can we write this down in the comment here that we're using a third account here because we can't use the same account we used to create pool?

Reply:
of course, I expanded the existing comment in my recent commit!